GraphQL Client: Add Apollo and use on profile settings page

umpox opened the pull request:
This PR:

- Adds [Apollo Client](https://www.apollographql.com/docs/react/).
- ~~Adds `typescript-react-apollo` to automatically generated typed `query` and `mutation` hooks.~~ https://github.com/sourcegraph/sourcegraph/pull/22380
- ~~Adds support and encourages writing queries in `.graphql` files, relying on generated code for initial logic~~ https://github.com/sourcegraph/sourcegraph/pull/22380
- Updates `EditUserProfileForm` to fire mutations through Apollo and respond accordingly. 
- Updates `UserSettingsProfilePage` to fire queries through Apollo and demonstrate how response data can be shared across components. Also adds tests for this page as an example for mocking different requests. 
- Adds and updates our web 'Working with GraphQL' documentation.

**This PR is structured into specific commits to make reviewing easier**

closes https://github.com/sourcegraph/sourcegraph/issues/22150
closes https://github.com/sourcegraph/sourcegraph/issues/22151

valerybugakov commented on client/web/src/user/settings/profile/UserSettingsProfilePage.test.tsx:
```diff
@@ -0,0 +1,111 @@
+    avatarURL: 'https://example.com/other-image.jpg',
+}
+
+const mocks: readonly MockedResponse[] = [
```
It would be cool to look into tooling around automatically generating mocks based on the GraphQL query or Typescript types.

Here's the list of interesting libraries that might be helpful:
https://www.npmjs.com/package/graphql-codegen-typescript-mock-data
https://www.npmjs.com/package/ts-auto-mock
https://github.com/marchaos/jest-mock-extended

umpox replied:
Yeah this is an interesting topic, I'll create an issue to investigate this as there's loads we can do here :)

valerybugakov commented on client/web/src/user/settings/profile/UserSettingsProfilePage.test.tsx:
```diff
@@ -0,0 +1,111 @@
+import { UPDATE_USER } from './EditUserProfileForm'
+import { UserSettingsProfilePage } from './UserSettingsProfilePage'
+
+jest.mock('../../../tracking/eventLogger', () => ({
+    eventLogger: { log: () => undefined, logViewEvent: () => undefined },
+}))
```
It seems that `eventLogger` is mocked in a lot of tests.
Does it make sense to mock it globally for all tests similar to `mockDate.js`?

umpox replied:
Yep, I'll do it in a different PR to keep this one clean

umpox replied:
https://github.com/sourcegraph/sourcegraph/pull/22313

valerybugakov commented on client/web/src/user/settings/profile/UserSettingsProfilePage.test.tsx:
```diff
@@ -0,0 +1,111 @@
+    avatarURL: 'https://example.com/image.jpg',
+    viewerCanChangeUsername: true,
+    createdAt: new Date().toISOString(),
+} as NonNullable<UserAreaUserFields>
```
Does it fail because of the required variables if the type is set without `as`? Then [jest-mock-extended](https://www.npmjs.com/package/graphql-codegen-typescript-mock-data) might be helpful here.

```ts
const mockUser: NonNullable<UserAreaUserFields> = { ... }

valerybugakov commented on client/web/src/user/settings/profile/UserSettingsProfilePage.test.tsx:
```diff
@@ -0,0 +1,111 @@
+            expect(queries.getByText('User profile updated.')).toBeVisible()
+            expect(usernameField).toHaveValue(newUserValues.username)
+            expect(displayNameField).toHaveValue(newUserValues.displayName)
+            expect(avatarURLField).toHaveValue(newUserValues.avatarURL)
```
🚀

valerybugakov commented on doc/dev/background-information/web/graphql.md:
```diff
@@ -1,49 +1,176 @@
-The preferred way to get a typed result from a GraphQL query or mutation is to pass the auto-generated interface for the query result and variables as type parameters to `requestGraphQL()`:
+We use `gql` template strings to declare our GraphQL queries.
+
+Each query must have a globally unique name. This makes sure the query gets picked up by our type generation and you receive autocompletion, syntax highlighting, hover tooltips and validation.
```
It would nice to add some guidelines on the preferred naming pattern to ensure the uniqueness of the query name. Not sure how it's structured right now, is there any pattern already used across the codebase?

valerybugakov commented on doc/dev/background-information/web/graphql.md:
```diff
@@ -1,49 +1,176 @@
+
+**Running your query**
+
+Once you have built your query, `codegen` will generate the correct request and response types. It will also create typed React hooks. Using a newly-generated hook, we can easily fire a request and handle the response correctly.
```
Does it make sense to add a link to the `codegen` command or one sentence from where it's coming from? I've seen that it's confusing, especially for teammates not aware of our `yarn watch-generate` command.

valerybugakov commented on doc/dev/background-information/web/graphql.md:
````diff
@@ -1,49 +1,176 @@
+}
+```
+
+## Frequently asked questions
+### How do I use the Apollo cache?
````
Nit: [markdownlint](https://marketplace.visualstudio.com/items?itemName=DavidAnson.vscode-markdownlint) usually complains about blank lines around headings: `Headings should be surrounded by blank lines`. It's not enabled for our `.md` files, but it looks like a widespread practice.

valerybugakov commented on client/web/src/backend/graphql.ts:
```diff
@@ -60,3 +61,14 @@ export const mutateGraphQL = (request: string, variables?: {}): Observable<Graph
+
+const uri = '/.api/graphql'
+
+export const client = new ApolloClient({
```
Do we want to share `client` with browser extension in the future? 
Maybe it makes sense to put it next to `requestGraphQLCommon` in `client/shared/src/graphql/graphql.ts`?

Then we could also share the `uri` constant with the `requestGraphQLCommon` implementation.

valerybugakov commented on doc/dev/background-information/web/graphql.md:
```diff
@@ -1,49 +1,176 @@
+  // useGetUserDisplayName is automatically typed with correct variables and response data
+  const { data, loading, error } = useGetUserDisplayName({ variables: { username }});
+
+  if (loading) {
+    // handle loading state
+  }
+
+  if (error) {
+    // handle error state
+  }
+
+  if (data) {
+    // display to user
+  }
```
Recently we [discussed](https://github.com/sourcegraph/sourcegraph/pull/22139#discussion_r653329771) the pattern currently used in our codebase with @vovakulikov. Curious, do generated hooks disallow [impossible](https://docs.sourcegraph.com/dev/background-information/languages/typescript#making-invalid-states-impossible-through-union-types) return states? 

If it's not the case it would nice to patch the interface of the generated hooks to disallow them on the type level.

5h1rU replied:
not sure 100% but AFAIK Apollo avoids this behavior by contract.

umpox replied:
I'd be hesitant to patch these types unless we know for sure that these are actually impossible. E.g., it is valid in the GQL spec to return `data` and `error` at the same time. Apollo might also support loading as true when data is also true, on refetch/polling data? Not sure. I'd be surprised if there wasn't a good reason for this in Apollo, but I'm struggling to find extra context in their repos. WDYT?

valerybugakov replied:
Sorry, I didn't mean to suggest patching only the return types. 

I was thinking of improving the `useQuery` hook itself to explicitly disallow the impossible states. It should be possible to provide a custom wrapper through the `graphql-codegen` config for some Apollo hooks that will be used in the generated code.

If there're no obvious reasons to patch it for now — that's awesome. We can review it later once we will have more context and usage examples in our repo. I just wanted to ensure that the mentioned discussion and the doc about impossible states are on our radar for GraphQL client implementation.

valerybugakov reviewed (approved):
Looks awesome! 
Looking forward to trying it out myself in the codebase. Thanks, @umpox 💪

5h1rU commented on doc/dev/background-information/web/graphql.md:
```diff
@@ -1,49 +1,176 @@
+
+const MyComponent = ({ username }: { username: string }) => {
+  // useGetUserDisplayName is automatically typed with correct variables and response data
+  const { data, loading, error } = useGetUserDisplayName({ variables: { username }});
```
Can you describe the scenario when the data coming from graphql needs to be handled to expose a different structure?

Sometimes we need to show some data structures depending on different factors, but we don't want to make a different request. Using a custom hook that uses the apollo autogenerated hook and then exposes the custom hook in the component with "reshaped" data would work.

```js
const useGetUserDisplayFullName = ({ variables: {username} }) => {
  const { data, loading, error } = useGetUserDisplayName({ variables: { username }});
  const fullName = `${data.firstname} ${data.lastname}`
  const newData = { ...data, fullName}
  return { newData, loading, error }
}

// MyComponent.tsx
const MyComponent = ({ username }: { username: string }) => {
  const { customData, loading, error } = useGetUserDisplayFullName({ variables: { username }})
  ...
}

valerybugakov commented on doc/dev/background-information/web/graphql.md:
```diff
@@ -1,67 +1,211 @@
-The preferred way to get a typed result from a GraphQL query or mutation is to pass the auto-generated interface for the query result and variables as type parameters to `requestGraphQL()`:
+We use `gql` template strings to declare our GraphQL queries.
+
+Each query must have a globally unique name as per the [GraphQL specification](https://spec.graphql.org/June2018/#sec-Operation-Name-Uniqueness). Typically we should name our queries similarly to how we might name a function, by describing what the query should do. This should also mean prefixing the name with an action, like `Get` or `Update`, this will help avoid collisions between queries and  ...
```
Thanks for adding this ⛑

eseliger replied:
`Get` and `Update` feel pretty redundant to me (see above), I think a proper verb in the mutation will always be enough to make the distinction. Ie `DeleteUser`, `TriggerEmail`, ... while queries would usually be something like `UserAreaProfile` or `UserEmails`. WDYT?

umpox replied:
Yeah I think that's a good assessment, will update the docs and examples!

valerybugakov reviewed (approved):
Shit it 🛳

umpox commented:
@5h1rU @valerybugakov Thanks for the reviews, I've updated this slightly after comments and improved the documentation.

I've also changed this to declare queries in `.graphql` files. Initially this was to avoid duplicating queries in both the original and generated `.ts` files when generating code. There's also some additional benefits:
- Better separation of concerns, and easier to find related queries by scanning the codebase
- Better editor support in `.graphql` files
- We can be more consistent by only declaring the query in a `.graphql` file and relying fully on generated code for basic request logic, will make refactoring easier in the future.

umpox commented on client/web/src/user/settings/cloud-ga.ts:
```diff
@@ -1,13 +1,23 @@
     authenticatedUser: Pick<AuthenticatedUser, 'id' | 'tags'>
 }
 
+export const requiredCloudGAUserFields = gql`
+    fragment RequiredGAUserFields on User {
+        id
+        tags
+        builtinAuth
+    }
+`
```
Few notes here:

Types that use directives `@include` or `@skip` now are correctly typed with `Maybe<>`, which is great, but this raised a few issues.

Firstly:
1. We rely on the `UserAreaUserFields` fragment in a lot of places, but this isn't really suited to be used like this. `tags` here is only included in the fragment when the user is a site admin. Relying on GQL fragments to generate types feels wrong to me, I wonder if anyone has a better solution here. Maybe export `User` as a type and just use `Pick`?
2. There is a bug in graphql-codegenerator that, when a directive causes the type to become `<Maybe>`, the `<Array>` type is stripped away. This isn't a problem for us right now and I've raised an issue against their code for this

umpox commented on client/web/src/user/area/UserArea.graphql:
```diff
@@ -0,0 +1,35 @@
+            name
+        }
+    }
+    # TODO: Re-enable when directive types are fixed in GraphQL code generator: https://github.com/sourcegraph/sourcegraph/issues/22338
+    tags # @include(if: $siteAdmin)
```
Note: I commented out the `@include` here. These changes actually now respond to `@include` and `@skip` directives by adding making these types nullable. The downside to this is that there is currently a bug in GraphQL code generator that also strips away the `Array<>` type when this happens. It has been fixed for object types but not primitive ones like `String[]`. I have reproduced and raised a bug. Will re-enable when fixed (this is the only case in the codebase and doing this has minimal impact)

felixfbecker replied:
Does this change not break anything? I could imagine that the reason for the `@include` is that this field will throw an error if queried as a non-site-admin, so it may break the page now if accessed by a non-side-admin?

eseliger replied:
Yes this will fail the query with an error, because only site admins may view this property. An older version of the generator didn't have this bug, maybe we can continue to use the old version?

umpox replied:
Ah I ran the query as a normal user and it didn't error so I assumed it was OK. Checking the code I now see that it is if the user is a site-admin *or* the same as user as the one being requested.

Hmm, frustrating. @eseliger You are right that there is a similar bug in codegen that has been fixed before. This one is slightly different and I've [reproduced](https://codesandbox.io/s/fervent-dijkstra-ub9hg?file=/document.graphql) it with the latest dependencies. I think `typescript-react-apollo` might be bumping a peer dependency which causes this (previously our types assumed `tags` was always defined so this was never an issue)

eseliger replied:
Hmm can you try to hard pin the dependency to `~1.17.0`? I think it was introduced in `1.18`.

felixfbecker commented:
Cool! 😎 

With using `.graphql` files, what's the way fragments can be shared across multiple queries? In TS, we would export/import them and inline them with a template string. Maybe you could convert a component that does that as an example/validation?

Also curious what the "better editor support" is you mentioned in Slack (we had autocompletion and validation before with a tsserver plugin and in CI – curious how does this work now? And are all needed extensions added to `.vscode/extensions.json` and do the CI checks still work?)

eseliger commented on client/web/src/user/area/UserArea.graphql:
```diff
@@ -0,0 +1,35 @@
+    tags # @include(if: $siteAdmin)
+}
+
+query GetUserProfile($username: String!, $siteAdmin: Boolean!) {
```
I find the prefix `Get` is a bit redundant. A `query` is always "getting" data, no? Maybe it can be a bit more specific to where it's being used instead. Ie. `UserAreaUserProfile` or something like that.

eseliger commented on client/web/src/user/area/UserArea.graphql:
```diff
@@ -0,0 +1,35 @@
+# GraphQL fragment for the User fields needed by UserArea.
```
Are the .graphql documents checked with our usual tooling, too? Inside TS code, we ran a couple of checks over them iirc.

umpox replied:
Trying to find these checks, do you know where they are? We get usual type checking, e.g. using an incorrect field. Unsure if there are other checks

eseliger replied:
Yeah I think that's it. I think it is a side-effect of running the generator(?). We also have a schema validator but I think that one doesn't work for queries/mutations, so probably all good here.

eseliger commented on client/web/src/user/area/UserArea.graphql:
```diff
@@ -0,0 +1,35 @@
+# GraphQL fragment for the User fields needed by UserArea.
```
Can this fragment still be shared?

umpox replied:
It can, essentially as a global variable, you don't get any code completion as far as I can tell though :(. I'm going to investigate this and a few different options in https://github.com/sourcegraph/sourcegraph/pull/22380 once this PR goes through

eseliger commented on doc/dev/background-information/web/graphql.md:
```diff
@@ -1,67 +1,211 @@
 
-## Writing a type-safe GraphQL query
+## GraphQL Client
+We use [Apollo Client](https://www.apollographql.com/docs/react/) to manage data-fetching and caching within our app. It provides a set of declarative interfaces which abstract away a lot of repeated code, whilst supporting a 'stale-while-revalidate' strategy though a normalized client cache of requested data.
```
```suggestion
We use [Apollo Client](https://www.apollographql.com/docs/react/) to manage data-fetching and caching within our app. It provides a set of declarative interfaces which abstract away a lot of repeated code, whilst supporting a 'stale-while-revalidate' strategy through a normalized client cache of requested data.
```

eseliger commented on doc/dev/background-information/web/graphql.md:
```diff
@@ -1,67 +1,211 @@
-        }
+const MyComponent = ({ username }: { username: string }) => {
+  // useGetUserDisplayName is automatically typed with correct variables and response data
+  const { data, loading, error } = useGetUserDisplayName({ variables: { username }});
```
```suggestion
  const { data, loading, error } = useGetUserDisplayName({ variables: { username } });
```

eseliger commented on doc/dev/background-information/web/graphql.md:
````diff
@@ -1,67 +1,211 @@
+
+```graphql
+# ./MyComponent.graphql
+query GetUserDisplayName($username: String!) {
+    user(username: $username) {
+        displayName
+    }
+}
````
Considering the section below on caching, should this query also return `__typename` and `id`?

umpox replied:
`id` yes :D `__typename` is appended automatically though :)

eseliger commented on client/shared/dev/generateGraphQlOperations.js:
```diff
@@ -79,6 +84,7 @@ async function generateGraphQlOperations() {
             noExport: false,
             enumValues: '@sourcegraph/shared/src/graphql-operations',
             interfaceNameForOperations: 'BrowserGraphQlOperations',
+            pureMagicComment: true,
```
maybe add this as a code comment, too.

eseliger commented on doc/dev/background-information/web/graphql.md:
```diff
@@ -1,67 +1,211 @@
 
+Other code uses `requestGraphQL()`, this is an improved approach which provides better types, but it doesn't scale well when requesting data across multiple areas of the application, and often leads to cross-component dependencies.
+
+Our goal is to migrate more code to use Apollo. This should make our components more self-contained, increase perceived performance with client-side caching and make it easier to write effective tests.
```
should we encourage people to not use the first two approaches in any new code?

eseliger replied:
Makes sense 👍

eseliger commented on client/web/src/user/area/UserArea.tsx:
```diff
@@ -193,15 +121,14 @@ export const UserArea: React.FunctionComponent<UserAreaProps> = ({
     )
 
-    if (user === undefined) {
+    if (!user) {
         return null // loading
     }
```
Shouldn't this use `loading` which is returned from the hook? 
And have error handling?
(Sorry for being picky, reference implementations are copied a lot of times)

felixfbecker replied:
It's a bit unfortunate that handling of the loading and error states is no longer _enforced_ through type-safety as before, potentially leading to exactly these accidental oversights.

eseliger replied:
True, I liked that as well.

eseliger commented on client/web/src/user/area/UserArea.tsx:
```diff
@@ -1,89 +1,30 @@
-    ).pipe(
-        map(dataOrThrowErrors),
-        map(data => {
-            if (!data.user) {
-                throw new Error(`User not found: ${JSON.stringify(args.username)}`)
-            }
```
This is not considered anymore afaict, and will result in an endless loading state. Should return a 404 page if not loading, no error and data.user is null.

umpox replied:
Thanks missed this, will re-add

eseliger commented on client/web/src/user/area/UserArea.tsx:
```diff
@@ -137,9 +78,6 @@ export interface UserAreaRouteContext
      */
     user: UserAreaUserFields
 
-    /** Called when the user is updated, with the full new user data. */
-    onUserUpdate: (newUser: UserAreaUserFields) => void
```
This is not required anymore, because the user is always kept up to date with the apollo cache, right? Just trying to validate my understanding 😬

umpox replied:
That's correct, cache chances will cause this component to re-render with the latest data

eseliger replied:
That's pretty neat

eseliger commented on client/web/src/user/settings/profile/UserSettingsProfilePage.test.tsx:
```diff
@@ -0,0 +1,110 @@
+
+import { UserSettingsProfilePage } from './UserSettingsProfilePage'
+
+jest.mock('../../../tracking/eventLogger', () => ({
+    eventLogger: { log: () => undefined, logViewEvent: () => undefined },
+}))
```
can be removed now :)

eseliger commented on client/web/src/user/settings/profile/UserSettingsProfilePage.tsx:
```diff
@@ -1,65 +1,21 @@
-
-            parentOnUpdate(newValue)
-
-            // In case the edited user is the current user, immediately reflect the changes in the
-            // UI.
-            refreshAuthenticatedUser()
-                .toPromise()
-                .finally(() => {})
```
This is still required, no? `authenticatedUser` doesn't use an apollo hook afaict.

eseliger commented on client/web/src/user/settings/profile/EditUserProfileForm.tsx:
```diff
@@ -1,113 +1,73 @@
-                        map(data => data.updateUser)
-                    )
-                    .toPromise()
-                eventLogger.log('UserProfileUpdated')
```
these two eventlogger calls aren't  required anymore?

umpox replied:
Thanks, not sure how I missed these 🤦

felixfbecker commented on client/web/src/user/settings/profile/UserSettingsProfilePage.test.tsx:
```diff
@@ -0,0 +1,110 @@
+    avatarURL: 'https://example.com/image.jpg',
+    viewerCanChangeUsername: true,
+    createdAt: new Date().toISOString(),
+} as UserAreaUserFields
```
Why is a cast needed (and correct/okay) here?
https://docs.sourcegraph.com/dev/background-information/testing_web_code#treat-tests-with-the-same-standards-to-type-safety-as-application-code

umpox commented:
FYI: There's still some things to iron out with the type enhancements, a few different ways of doing things. I'm moving that to a separate PR to keep this one focused

umpox commented on package.json:
```diff
@@ -96,9 +96,10 @@
+    "@graphql-codegen/cli": "^1.21.5",
+    "@graphql-codegen/typescript": "1.22.3",
     "@graphql-codegen/typescript-operations": "1.17.8",
+    "@graphql-codegen/typescript-react-apollo": "^2.2.7",
```
Going to address in https://github.com/sourcegraph/sourcegraph/pull/22380

umpox commented:
Have made some changes. this should be ready for re-review now (I think). Some notes:

- Typescript enhancements will be done in a different PR to keep this one smaller
- Apollo expects `DocumentNode` as a query type. I've created a small wrapper so we can keep supporting `string` queries for now. Want to consider refactoring to use DocumentNode or [TypedDocumentNode](https://the-guild.dev/blog/typed-document-node) as part of TS enhancements 

@felixfbecker The `.graphql` fragments can be shared across other fragments globally, codegen handles the output and replicates injecting them into template strings. That being said, this approach isn't perfect so I'm going to revisit this in a different PR.

In terms of testing, I added a section in the docs for this. We can mock queries/mutations globally using `MockedProvider` and we don't have to worry about dependency injection

eseliger reviewed (approved):
🌟 
:shipit:

felixfbecker commented on babel.config.js:
```diff
@@ -36,8 +36,6 @@ module.exports = api => {
             'web.url-search-params',
             // Commonly needed by extensions (used by vscode-jsonrpc)
             'web.immediate',
-            // Avoids issues with RxJS interop
-            'esnext.symbol.observable',
```
How come this is not needed?

The last time I looked into this, I remember that having an explicit `import` didn't actually do anything unless it was listed here, because Babel replaces (expands) any import to only the list of polyfills needed and listed here in `include`. Which is why you also don't see `web.url` etc listed out in polyfills.ts, they are added by Babel automatically, because they are here in `include`.

It's possible to verify this by using the `babel` CLI and giving it the `polyfill.ts` file path directly, it will print the compiled file, which shows the expanded imports.

umpox replied:
```
npx babel client/shared/src/polyfills/polyfill.ts 

import "core-js/modules/es.aggregate-error.js";
import "core-js/modules/es.promise.any.js";
import "core-js/modules/es.promise.finally.js";
import "core-js/modules/es.reflect.to-string-tag.js";
import "core-js/modules/es.string.match-all.js";
import "core-js/modules/es.string.replace.js";
import "core-js/modules/es.string.replace-all.js";
import "core-js/modules/es.typed-array.float32-array.js";
import "core-js/modules/es.typed-array.float64-array.js";
import "core-js/modules/es.typed-array.int8-array.js";
import "core-js/modules/es.typed-array.int16-array.js";
import "core-js/modules/es.typed-array.int32-array.js";
import "core-js/modules/es.typed-array.uint8-array.js";
import "core-js/modules/es.typed-array.uint8-clamped-array.js";
import "core-js/modules/es.typed-array.uint16-array.js";
import "core-js/modules/es.typed-array.uint32-array.js";
import "core-js/modules/es.typed-array.from.js";
import "core-js/modules/es.typed-array.of.js";
import "core-js/modules/web.dom-collections.iterator.js";
import "core-js/modules/web.immediate.js";
import "core-js/modules/web.url.js";
import "core-js/modules/web.url.to-json.js";
import "core-js/modules/web.url-search-params.js";
import "core-js/modules/esnext.symbol.observable.js";
```

My understanding is that Babel will expand the imports to only import polyfills needed against our `browserslist`. Then we can use `include` to force Babel to _always_ include a certain polyfill. I don't feel like we need to force include this here, WDYT?

felixfbecker replied:
Ah, okay, so you just removed this because it is not supported by all of our browsers anyway and implicitly included. I guess I just wanted to be _extra_ sure it's included and have some place to document the context in a short comment.

With what you wrote, is the explicit `import` you added in this commit too necessary though?

umpox replied:
I believe so, because the polyfill isn't included in `core-js/stable` so we have to be explicit.

Given how much of a pain this was to debug, I'm in favor of keeping it in `include` and to be extra sure we have it :D

felixfbecker replied:
Ah I see. That would be really great to document with a comment above the `import`!

courier-new commented on client/shared/src/graphql/graphql.ts:
```diff
@@ -69,11 +87,71 @@ export function requestGraphQLCommon<T, V = object>({
+ * @param options Operation variables and request configuration
+ * @returns GraphQL response
+ */
+export function useQuery<TData = any, TVariables = OperationVariables>(
+    query: RequestDocument,
+    options: QueryHookOptions<TData, TVariables>
+): QueryResult<TData, TVariables> {
+    const documentNode = useDocumentNode(query)
+    return useApolloQuery(documentNode, options)
+}
+
+/**
+ * Send a mutation to GraphQL and respond to updates.
+ * Wrapper around Apollo `useMutation` that supports `DocumentNode` and `string` types.
+ *
+ * @param mutation GraphQL operation payload.
+ * @param options Operation variables and request configuration
+ * @returns GraphQL response
+ */
+export function useMutation<TData = any, TVariables = OperationVariables>(
+    mutation: RequestDocument,
```
So I'm a bit late to this party, but I wanted to make one strong suggestion based on some past experience, which is to set the default type for generic argument `TData` to `unknown` rather than `any`, for both `useQuery` and `useMutation`.

I know Apollo's own methods that we are wrapping use `TData = any` in their definitions, but from personal experience, this **always** leads to leaking `any`s and then, inevitably, uncaught type errors. Let me demonstrate:

So you and I (and likely many of the reviewers on this PR) know that in order to ensure type safety on a call to one of these hooks, we need to narrow `TData` to something more specific than `any`. The most obvious way to do this it to pass the types when calling the function, like you do in `UserArea.tsx`:

```tsx
const { data } = useQuery<UserAreaUserProfileResult, UserAreaUserProfileVariables>(USER_AREA_USER_PROFILE, ...);

const userID = data?.user?.id; // I'm safe!
```

✅ Here, the type of `data` is `UserAreaUserProfileResult | undefined`, and the type of `userID` is `string | undefined`, like we want.

The other way would be if we're able to infer the type from the `QueryHookOptions` or `MutationHookOptions`, like this:

```tsx
const { data } = useQuery(
  USER_AREA_USER_PROFILE,
  { onCompleted: (data: UserAreaUserProfileResult) => { ... } }
);

const userID = data?.user?.id; // I'm safe!
```

✅ Again, the type of `data` is properly inferred to be `UserAreaUserProfileResult | undefined`, and the type of `userID` is `string | undefined`, like we want.

However, notice what happens when we have neither of these things:

```tsx
const { data } = useQuery(USER_AREA_USER_PROFILE, ...);

const userID = data?.user?.id; // Who knows what I am!
const badUserID = data.user.id; // 🚫 Uncaught TypeError: Cannot read property 'id' of undefined
```

❌ Here, both `data` and `userID` get the type `any`, and the TS compiler loses the ability to catch potential issues.

courier-new commented on client/shared/src/graphql/graphql.ts:
```diff
@@ -69,11 +87,71 @@ export function requestGraphQLCommon<T, V = object>({
+ * @param options Operation variables and request configuration
+ * @returns GraphQL response
+ */
+export function useQuery<TData = any, TVariables = OperationVariables>(
+    query: RequestDocument,
+    options: QueryHookOptions<TData, TVariables>
+): QueryResult<TData, TVariables> {
+    const documentNode = useDocumentNode(query)
+    return useApolloQuery(documentNode, options)
+}
+
+/**
+ * Send a mutation to GraphQL and respond to updates.
+ * Wrapper around Apollo `useMutation` that supports `DocumentNode` and `string` types.
+ *
+ * @param mutation GraphQL operation payload.
+ * @param options Operation variables and request configuration
+ * @returns GraphQL response
+ */
+export function useMutation<TData = any, TVariables = OperationVariables>(
```
So I'm a bit late to this party, but I wanted to make one strong suggestion based on some past experience, which is to set the default type for generic argument `TData` to `unknown` rather than `any`, for both `useQuery` and `useMutation`.

I know Apollo's own methods that we are wrapping use `TData = any` in their definitions, but from personal experience, this **always** leads to leaking `any`s and then, inevitably, uncaught type errors. Let me demonstrate:

So you and I (and likely many of the reviewers on this PR) know that in order to ensure type safety on a call to one of these hooks, we need to narrow `TData` to something more specific than `any`. The most obvious way to do this it to pass the types when calling the function, like you do in `UserArea.tsx`:

```tsx
const { data } = useQuery<UserAreaUserProfileResult, UserAreaUserProfileVariables>(USER_AREA_USER_PROFILE, ...);

const userID = data?.user?.id; // I'm safe!
```

✅ Here, the type of `data` is `UserAreaUserProfileResult | undefined`, and the type of `userID` is `string | undefined`, like we want.

The other way would be if we're able to infer the type from the `QueryHookOptions` or `MutationHookOptions`, like this:

```tsx
const { data } = useQuery(
  USER_AREA_USER_PROFILE,
  { onCompleted: (data: UserAreaUserProfileResult) => { ... } }
);

const userID = data?.user?.id; // I'm safe!
```

✅ Again, the type of `data` is properly inferred to be `UserAreaUserProfileResult | undefined`, and the type of `userID` is `string | undefined`, like we want.

However, notice what happens when we have neither of these things:

```tsx
const { data } = useQuery(USER_AREA_USER_PROFILE, ...);

const userID = data?.user?.id; // Who knows what I am!
const badUserID = data.user.id; // 🚫 Uncaught TypeError: Cannot read property 'id' of undefined
```

❌ Here, both `data` and `userID` get the type `any`, and the TS compiler loses the ability to catch potential issues.

We might notice lots of ESLint warnings pop up if we tried to do this, like:
- Unsafe array destructuring of a tuple element with an any value, or
- Unsafe assignment of an any value

However, as far as I know, ESLint warnings don't report as a failure on CI (and for good reason -- `yarn run eslint` in `web` currently seems to produce about 675 warnings, many of which are about `any` values 😅).

Notably, this does **not** produce any TSC errors.

Needing to pass generic arguments or explicitly type the arguments on an `options` object function are very much un-obvious requirements, especially for someone who is newer to TS or to frontend development more generally, and it shouldn't be the sole responsibility of reviewers to know and catch this when it's missed. But dealing with `any`s anywhere that network response data is involved is an absolute nightmare, so it's not something we can afford to be careless with.

Now let me share what happens if we change `TData = any` to `TData = unknown`, instead, with the same example query:

```tsx
const { data } = useQuery(USER_AREA_USER_PROFILE, ...);

const userID = data?.user?.id; // 🚫 TS compiler error! Object is of type 'unknown'.
```

✅ Here, `data` gets assigned the type `unknown`. Without knowing what type `data` is, the TS compiler will now adamantly refuse to let you access properties on it and throw a big ole compiler error when you try. No longer just some linter warnings, this is an immediate red flag to developers that something else needs to happen here and a CI failure when it comes time to submitting code for review.

And, if for some reason you really need to treat `data` as `any` for a particular request, you still can, you just have to explicitly say so:

```tsx
const { data } = useQuery<any>(USER_AREA_USER_PROFILE, ...);
```

Arguably we would want to take a similar approach to `TVariables`, but it's way less likely to manifest as a browser exception, and it'll be pretty obvious when the variable type is wrong to developers because the query/mutation simply won't work.

courier-new replied:
(I also know you're actively working on a follow-up PR with the codegen stuff so please lmk if you'd prefer I move this comment over there instead!!)

umpox replied:
Yep will do this, also considering implementing [TypedDocumentNode](https://the-guild.dev/blog/typed-document-node) and enforcing that to try to make this scenario easier to avoid in the future.

Thanks for these comments, appreciate the 👀 !

courier-new replied:
Oh that looks awesome! I'm familiar with their generative tools so I'd certainly be interested in seeing how this next evolution works. Thanks for sharing, and happy the comments are helpful. 🙂

courier-new commented on doc/dev/background-information/web/graphql.md:
````diff
@@ -1,49 +1,195 @@
+
+## Frequently asked questions
+
+### How do I use the Apollo cache?
+Apollo uses a normalized in-memory cache to store the results of different queries, most of this happens automatically!
+
+Apollo generates a composite key for each identifiable object in a response. This is typically done by combining the `__typeName` with the `id` that is returned in the response. For example:
+
+When firing this request:
+```
+{
+  user {
+    __typename
+    id
+    displayName
+  }
+}
+```
+
+Assuming `__typename === User` and `id === 1234`, the response data is added to the cache under the key:
+`User:1234`.
+
+When a different query requests similar data, Apollo will merge both responses into the cache and update both parts of the UI with the latest data. This is useful to form links between components whilst still ensuring that the components are self-contained. For example, if a mutation updated `displayName` in a different component, then these components that *query* `displayName` would automatical ...
+
+**All queries should return an object identifier**. If this identifier is not under the `id` field, we need to inform Apollo which field to use in order to generate the correct key. See the [docs](https://www.apollographql.com/docs/react/caching/cache-configuration/#customizing-identifier-generation-by-type) for more information on this.
+
````
One other note, obviously having access to the Apollo cache has a ton of benefits and is one of the biggest reasons to adopt it or a similar GraphQL client library in the first place, but I think one thing that is missing from your very helpful docs here is knowing when *not* to use the cache. Or, in practice for Apollo, how to bypass it.

The stale-while-revalidate strategy is very widely used these days, and with good reason, but it's not the best strategy all the time. When a user visits a page with data that's in the cache, they'll see that data right away -- but it **may be out-of-date**. Of course, Apollo will be refetching that data in the background, but there will be "flash" of old data, if something changed, which can be a bit jarring if the user is actively looking at that part of the UI. It's important to consider when this cached data may actually be unwanted, and just a plain loading spinner preferred. Some (not-necessarily-Sourcegraph-related) examples that come to mind:

- One-off tokens
- Random data
- Data tied to the current date/time
- Alerts/notifications
- Shared data that may be impacted by other users' actions
- When a mutation has indirect, async side effects on other data

Not to say that it's a rule that you shouldn't use the cache if any of these conditions are true, just that it's worth considering on a case-by-case basis what to do when you have stale data, and if you should just show it, show it _with_ a loading spinner, or hide it and only show a loading spinner.

For developers coming from the existing setup with no caching at all, and who are regularly developing from the perspective of a user's first load (refreshing the page, using storybooks, generally isolated development, etc.), it's not always obvious when stale data in the cache could cause worse UX. From personal experience, many developers **won't** think about it until after it's bitten them. IMO Apollo's docs don't really do a good job of discussing this, either, and the details on how to do it are a bit hard to find unless you already know it's called [fetchPolicy](https://www.apollographql.com/docs/react/data/queries/#setting-a-fetch-policy). 😅

I'm happy to rephrase this a bit and submit a PR for this myself, or let you take/paraphrase this however you see fit! Just wanted to make sure to raise this, as it's another recurring issue I've seen happen when it comes to client-side caching tools that's a bit harder to guide through tooling and code examples alone. 🙂

courier-new replied:
Will do!!
